### PR TITLE
[BugFix] Fix semi-type bugs

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

1. Fix equals binary with null check

`equals` method, add `equals_safe` parameter and update return value

before:
```
MySQL td> select map{'a': 1} = map{'a': null}
+--------------------------------+
| (map{'a':1}) = (map{'a':NULL}) |
+--------------------------------+
| false                          |
+--------------------------------+
```

now:
```
MySQL td> select map{'a': 1} = map{'a': null}
+--------------------------------+
| (map{'a':1}) = (map{'a':NULL}) |
+--------------------------------+
| <null>                         |
+--------------------------------+
```

2. Fix semi-type cast error message
3. Fix load struct crash: lose field names of destination type
4. Fix result of constant `named_struct` is error

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Review by kangkaisen — DISMISSED** [review]

File: be/src/column/column.h [hunk 1]
```diff
@@ -347,7 +347,9 @@ class Column {
     virtual int compare_at(size_t left, size_t right, const Column& rhs, int nan_direction_hint) const = 0;
 
     // For some columns equals will be overwritten for more efficient
-    virtual bool equals(size_t left, const Column& rhs, size_t right) const {
+    // When safe equals, 0: false, 1: true
+    // When unsafe equals, -1: NULL, 0: false, 1: true
+    virtual int equals(size_t left, const Column& rhs, size_t right, bool safe_eq = true) const {
```
Comment (fzhedu) [inline be/src/column/column.h:352]: how about int8?

File: be/src/column/array_column.cpp [hunk 2]
```diff
@@ -409,16 +409,17 @@ bool ArrayColumn::equals(size_t left, const Column& rhs, size_t right) const {
...
     if (lhs_end - lhs_offset != rhs_end - rhs_offset) {
-        return false;
+        return 0;
     }
     while (lhs_offset < lhs_end) {
-        if (!_elements->equals(lhs_offset, *(rhs_array._elements.get()), rhs_offset)) {
-            return false;
+        auto res = _elements->equals(lhs_offset, *(rhs_array._elements.get()), rhs_offset, safe_eq);
+        if (res != 1) {
+            return res;
```
Comment (fzhedu) [inline be/src/column/array_column.cpp:417]: should diff -1 and 0 ?

File: be/src/column/map_column.cpp [hunk 3]
```diff
@@ -425,23 +425,27 @@ bool MapColumn::equals(size_t left, const starrocks::Column& rhs, size_t right)
     for (uint32_t i = lhs_offset; i < lhs_end; ++i) {
         bool found = false;
         for (uint32_t j = rhs_offset; j < rhs_end; ++j) {
-            if (!_keys->equals(i, *(rhs_map._keys.get()), j)) {
+            int res = _keys->equals(i, *(rhs_map._keys.get()), j, safe_eq);
+            if (res == -1) {
+                return -1;
```
Comment (fzhedu) [inline be/src/column/map_column.cpp:430]: should not directly return
{null:1, 'a':2} == {null:2,'a':2} => false
{null:1, 'a':2} == {null:1, 'a':3} => false
{null:1, 'a':2} == {null:1, 'b':2} => false

File: be/src/column/struct_column.cpp [hunk 4]
```diff
@@ -291,29 +291,32 @@ size_t StructColumn::filter_range(const Filter& filter, size_t from, size_t to)
...
 
-    for (int i = 0; i < _fields.size(); ++i) {
+    auto lsize = _fields.size();
+    auto rsize = rhs_struct._fields.size();
+    auto size = std::min(lsize, rsize);
+
+    for (int i = 0; i < size; ++i) {
         auto cmp = _fields[i]->compare_at(left, right, *rhs_struct._fields[i].get(), nan_direction_hint);
         if (cmp != 0) {
             return cmp;
```
Comment (fzhedu) [inline be/src/column/struct_column.cpp:302]: should diff -1, 0

File: be/src/column/struct_column.cpp [hunk 5]
```diff
@@ -291,29 +291,32 @@ size_t StructColumn::filter_range(const Filter& filter, size_t from, size_t to)
...
     const auto& rhs_struct = down_cast<const StructColumn&>(rhs);
     if (_fields.size() != rhs_struct._fields.size()) {
         return false;
     }
     for (int i = 0; i < _fields.size(); ++i) {
-        if (!_fields[i]->equals(left, *rhs_struct._fields[i].get(), right)) {
-            return false;
+        auto res = _fields[i]->equals(left, *rhs_struct._fields[i].get(), right, safe_eq);
+        if (res != 1) {
+            return res;
```
Comment (fzhedu) [inline be/src/column/struct_column.cpp:316]: should diff 0, -1.
if 0, return 0,
if -1, should return after all check

File: be/src/column/array_column.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/column/array_column.cpp:417]: why ?

Comment (Seaven) [reply be/src/column/array_column.cpp:417]: emmm, you are right

File: be/src/column/column.h (hunk 1 above)
Comment (Seaven) [reply be/src/column/column.h:352]: no problem, just keep same as `compare_at`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Review by stdpain — DISMISSED** [review]

File: be/src/column/map_column.cpp [hunk 6]
```diff
@@ -422,26 +422,40 @@ bool MapColumn::equals(size_t left, const starrocks::Column& rhs, size_t right)
...
     }
 
+    bool has_null = false;
     for (uint32_t i = lhs_offset; i < lhs_end; ++i) {
         bool found = false;
         for (uint32_t j = rhs_offset; j < rhs_end; ++j) {
-            if (!_keys->equals(i, *(rhs_map._keys.get()), j)) {
+            int res = _keys->equals(i, *(rhs_map._keys.get()), j, safe_eq);
+            if (res == EQUALS_FALSE) {
                 continue;
```
Comment (fzhedu) [inline be/src/column/map_column.cpp:431]: why not directly return false?

File: be/src/column/map_column.cpp [hunk 7]
```diff
@@ -422,26 +422,40 @@ bool MapColumn::equals(size_t left, const starrocks::Column& rhs, size_t right)
...
             }
 
+            // key is null && isn't safe eq, don't need check value
+            has_null |= (res == EQUALS_NULL);
+            if (res == EQUALS_NULL && !safe_eq) {
+                found = true;
+                break;
+            }
```
Comment (fzhedu) [inline be/src/column/map_column.cpp:439]: {null:2} == {null:3} => null when !safe_eq?

File: be/src/column/struct_column.cpp [hunk 8]
```diff
@@ -291,29 +291,37 @@ size_t StructColumn::filter_range(const Filter& filter, size_t from, size_t to)
...
         return false;
     }
+
+    int ret = EQUALS_TRUE;
     for (int i = 0; i < _fields.size(); ++i) {
-        if (!_fields[i]->equals(left, *rhs_struct._fields[i].get(), right)) {
-            return false;
+        auto tmp = _fields[i]->equals(left, *rhs_struct._fields[i].get(), right, safe_eq);
+        if (tmp == 0) {
+            return 0;
```
Comment (fzhedu) [inline be/src/column/struct_column.cpp:318]: use _EQUALS_FALSE to keep consistent

File: be/src/column/map_column.cpp (hunk 7 above)
Comment (Seaven) [reply be/src/column/map_column.cpp:439]: yeah, I think the result is `NULL`. btw, trino don't support `NULL` as key

Comment (fzhedu) [reply be/src/column/map_column.cpp:439]: but I think it should return false

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

File: be/src/column/map_column.cpp (hunk 7 above)
Comment (Seaven) [reply be/src/column/map_column.cpp:439]: emmm, if return false that meanings key is safe equals?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|3|3|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]
LGTM

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**